File: copilot-token-budget/core/internal/budget/tracker.go
// Package budget converts raw nanoAIU billing values into credits, dollars, and
// budget-state summaries for the AT&T Copilot Token Budget tool.
package budget

// Billing unit conversions.
const (
	NanoAIUPerCredit        int64   = 1_000_000_000
	DollarsPerCredit        float64 = 0.01
	MonthlyAllowanceCredits int     = 7_000
)

// Sonnet token pricing (credits per million tokens).
const (
	SonnetInputRate  float64 = 300
	SonnetOutputRate float64 = 1_500
)

// BudgetStatus represents the health level of the current credit usage.
type BudgetStatus string

const (
	StatusOK       BudgetStatus = "OK"       // < 60% of allowance used
	StatusWarning  BudgetStatus = "WARNING"  // 60–90% of allowance used
	StatusCritical BudgetStatus = "CRITICAL" // > 90% of allowance used
)

// BudgetState holds a fully computed budget summary for a set of sessions.
//
// The json tags pin the wire shape to camelCase so it matches the TypeScript
// BudgetState exactly (note: RemainingCredit serializes as the plural
// "remainingCredits" to match the TS field). Without these tags Go would emit
// PascalCase and diverge from the extension's exports.
type BudgetState struct {
	UsedCredits     float64      `json:"usedCredits"`
	AllowedCredits  int          `json:"allowedCredits"`
	UsedPct         float64      `json:"usedPct"`
	RemainingCredit float64      `json:"remainingCredits"`
	Status          BudgetStatus `json:"status"`
}

// FromNanoAIU converts raw nanoAIU billing units to credits.
func FromNanoAIU(nanoAIU int64) float64 {
	return float64(nanoAIU) / float64(NanoAIUPerCredit)
}

// ToDollars converts credits to US dollars.
func ToDollars(credits float64) float64 {
	return credits * DollarsPerCredit
}

// Calculate sums nanoAIUValues, converts to credits, and computes the full BudgetState.
// If allowance is <= 0, MonthlyAllowanceCredits is used instead.
func Calculate(nanoAIUValues []int64, allowance int) BudgetState {
	if allowance <= 0 {
		allowance = MonthlyAllowanceCredits
	}

	var totalNano int64
	for _, v := range nanoAIUValues {
		totalNano += v
	}

	used := FromNanoAIU(totalNano)
	allowed := float64(allowance)
	pct := used / allowed * 100
	remaining := allowed - used

	return BudgetState{
		UsedCredits:     used,
		AllowedCredits:  allowance,
		UsedPct:         pct,
		RemainingCredit: remaining,
		Status:          statusFor(pct),
	}
}

// EstimateInstructionCostPerSession estimates the credit cost of always-loaded
// instruction file tokens across a typical 50-turn session.
//
// Formula: (totalTokens * 50 * SonnetInputRate) / 1_000_000
func EstimateInstructionCostPerSession(totalTokens int64) (credits float64, dollars float64) {
	const turnsPerSession = 50
	credits = (float64(totalTokens) * turnsPerSession * SonnetInputRate) / 1_000_000
	dollars = ToDollars(credits)
	return credits, dollars
}

// statusFor returns the BudgetStatus for a given usage percentage.
func statusFor(pct float64) BudgetStatus {
	switch {
	case pct > 90:
		return StatusCritical
	case pct >= 60:
		return StatusWarning
	default:
		return StatusOK
	}
}

File: copilot-token-budget/extension/src/budget/tracker.ts
// budget/tracker.ts — credit conversion and budget state calculation.
// TypeScript port of phase-1/session-manager/internal/budget/tracker.go.
// Zero npm runtime dependencies — pure arithmetic (ADR-003).

import { Session, BudgetState } from '../types';

// formatCreditsDisplay renders raw credits with thousands separators and up to two
// decimals — parity with the Go side (e.g. "8,554.03", "656.54"). Credits are already
// credits (nanoAIU / 1e9), so there is no further scaling and no "B"/billions unit.
function formatCreditsDisplay(credits: number): string {
  return credits.toLocaleString(undefined, { maximumFractionDigits: 2 });
}

// Billing unit conversions — mirrors Go named constants exactly.
export const NANO_AIU_PER_CREDIT = 1_000_000_000;
export const DOLLARS_PER_CREDIT  = 0.01;
export const MONTHLY_ALLOWANCE   = 7_000;

// Sonnet token pricing (credits per million tokens).
export const SONNET_INPUT_RATE  = 300;
export const SONNET_OUTPUT_RATE = 1_500;

// fromNanoAIU converts raw nanoAIU billing units to credits.
export function fromNanoAIU(nanoAIU: number): number {
  return nanoAIU / NANO_AIU_PER_CREDIT;
}

// toDollars converts credits to US dollars.
export function toDollars(credits: number): number {
  return credits * DOLLARS_PER_CREDIT;
}

// calculate sums nanoAIU from sessions, converts to credits, and returns a BudgetState.
// allowance defaults to MONTHLY_ALLOWANCE when not provided or <= 0.
export function calculate(sessions: Session[], allowance?: number): BudgetState {
  const allowed = (allowance != null && allowance > 0) ? allowance : MONTHLY_ALLOWANCE;

  const totalNano = sessions.reduce((sum, s) => sum + s.totalNanoAIU, 0);
  const used = fromNanoAIU(totalNano);
  const usedPct = (used / allowed) * 100;
  const remainingCredits = allowed - used;

  return {
    usedCredits: used,
    allowedCredits: allowed,
    usedPct,
    remainingCredits,
    status: statusFor(usedPct),
  };
}

// estimateInstructionCostPerSession estimates the credit cost of always-loaded
// instruction tokens across a typical 50-turn session.
// Formula: (totalTokens * 50 * SONNET_INPUT_RATE) / 1_000_000
export function estimateInstructionCostPerSession(totalTokens: number): { credits: number; dollars: number } {
  const TURNS_PER_SESSION = 50;
  const credits = (totalTokens * TURNS_PER_SESSION * SONNET_INPUT_RATE) / 1_000_000;
  return { credits, dollars: toDollars(credits) };
}

// statusBarText returns the VS Code status bar label for a BudgetState.
// Uses VS Code ThemeIcon syntax for colour hints:
//   CRITICAL → $(circle-filled)  WARNING → $(warning)  OK → $(check)
export function statusBarText(state: BudgetState): string {
  const icon = state.status === 'CRITICAL'
    ? '$(circle-filled) '
    : state.status === 'WARNING'
      ? '$(warning) '
      : '$(check) ';
  return `${icon}💰 ${formatCreditsDisplay(state.usedCredits)} / ${formatCreditsDisplay(state.allowedCredits)}`;
}

// statusFor maps a usage percentage to a BudgetState status value.
function statusFor(pct: number): BudgetState['status'] {
  if (pct > 90) {
    return 'CRITICAL';
  }
  if (pct >= 60) {
    return 'WARNING';
  }
  return 'OK';
}
